sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SideInputLoadTest.java

```diff
@@ -0,0 +1,243 @@
+        .apply("Collect end time metrics", ParDo.of(runtimeMonitor));
+  }
+
+  private void performTestWithHashMap(
+      PCollection<KV<byte[], byte[]>> input, Optional<SyntheticStep> syntheticStep) {
+    applyStepIfPresent(input, "Synthetic step", syntheticStep);
+    PCollectionView<Map<byte[], byte[]>> sideInput = input.apply(View.asMap());
+    input
+        .apply(ParDo.of(new SideInputTestWithHashMap(sideInput)).withSideInputs(sideInput))
+        .apply("Collect end time metrics", ParDo.of(runtimeMonitor));
+  }
+
+  private void performTestWithIterable(
+      PCollection<KV<byte[], byte[]>> input, Optional<SyntheticStep> syntheticStep) {
+    applyStepIfPresent(input, "Synthetic step", syntheticStep);
+    PCollectionView<Iterable<KV<byte[], byte[]>>> sideInput;
+    if (options.getWindowCount() != 0) {
+      long windowDurationMilis = sourceOptions.numRecords / options.getWindowCount();
+      sideInput =
+          input
+              .apply(Window.into(FixedWindows.of(Duration.millis(windowDurationMilis))))
+              .apply(View.asIterable());
+    } else {
+      sideInput = input.apply(View.asIterable());
+    }
+
+    input
+        .apply(ParDo.of(new SideInputTestWithIterable(sideInput)).withSideInputs(sideInput))
+        .apply("Collect end time metrics", ParDo.of(runtimeMonitor));
+  }
+
+  private static class AddTimestamps extends DoFn<KV<byte[], byte[]>, KV<byte[], byte[]>> {
+    private static Instant timestamp = TIME;
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      timestamp = TIME.plus(1L);
+      c.outputWithTimestamp(c.element(), timestamp);
+    }
+  }
+
+  private static class SideInputTestWithList extends DoFn<KV<byte[], byte[]>, KV<byte[], byte[]>> {
+
+    private final PCollectionView<List<KV<byte[], byte[]>>> sideInput;
+
+    public SideInputTestWithList(PCollectionView<List<KV<byte[], byte[]>>> sideInput) {
+      this.sideInput = sideInput;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      List<KV<byte[], byte[]>> si = c.sideInput(sideInput);
+
+      for (KV<byte[], byte[]> sideInputElement : si) {
+        // for every _input_ element iterate over all _sideInput_ elements
+        // count consumed bytes, examine memory usage, etc (Metrics API).
+        byte[] key = sideInputElement.getKey();
+      }
+    }
+  }
+
+  private class SideInputTestWithHashMap extends DoFn<KV<byte[], byte[]>, KV<byte[], byte[]>> {
+
+    private final PCollectionView<Map<byte[], byte[]>> sideInput;
+
+    public SideInputTestWithHashMap(PCollectionView<Map<byte[], byte[]>> sideInput) {
+      this.sideInput = sideInput;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      Map<byte[], byte[]> si = c.sideInput(sideInput);
+      // access Map elements according to options - % of random access
+      ArrayList<byte[]> keyList = new ArrayList<>(si.keySet());
```

Comment:
This part doesn't work well at high scale for some map implementations. An important use case is accessing just some keys and not others. A runner/SDK may have a separate place where they write the keySet so that you can make this call efficiently. But they may not. We don't want to make this test so sensitive to that (you could separate test efficiency of keySet). Can you calculate random keys in advance based on `sourceOptions`?

Reply:
I don't know if it's possible - I'll look into it

Reply:
I think calculating the keys beforehand would add a huge code overhead to the tests. I'm also still unsure how would I go about that. If you really think it would be worth it - can you give me some pointers regarding the approach to this issue?